plugins/net_plugin/net_plugin.cpp

```diff
@@ -2593,42 +2625,96 @@ namespace eosio {
-         time_point_sec expires{now.to_time_point() + my_impl->p2p_dedup_cache_expire_time_us};
+         time_point_sec expires{fc::time_point::now() + my_impl->p2p_dedup_cache_expire_time_us};
          expires = std::min( trx_expires, expires );
          local_txns.insert( node_transaction_state{
             .id = id,
             .expires = expires,
-            .connection_id = connection_id} );
+            .connection_ids = {c.connection_id},
+            .have_trx = true } );
+         ++c.trx_entries_size;
       }
-      return added;
+      return {c.trx_entries_size, already_have_trx};
+   }
+
+   size_t dispatch_manager::add_peer_txn_notice( const transaction_id_type& id, const time_point_sec& trx_expires, connection& c )
+   {
+      fc::lock_guard g( local_txns_mtx );
+
+      auto& id_idx = local_txns.get<by_id>();
+      if (auto tptr = id_idx.find( id ); tptr != id_idx.end()) {
+         if (tptr->connection_ids.insert(c.connection_id).second)
+            ++c.trx_entries_size;
+      } else {
+         // expire at either transaction expiration or configured max expire time whichever is less
+         time_point_sec expires{fc::time_point::now() + my_impl->p2p_dedup_cache_expire_time_us};
+         expires = std::min( trx_expires, expires );
+         local_txns.insert( node_transaction_state{
+            .id = id,
+            .expires = expires,
+            .connection_ids = {c.connection_id},
+            .have_trx = false } );
+         ++c.trx_entries_size;
+      }
+      return c.trx_entries_size;
    }
 
-   bool dispatch_manager::have_txn( const transaction_id_type& tid ) const {
+   connection_id_set dispatch_manager::peer_connections(const transaction_id_type& id) const {
       fc::lock_guard g( local_txns_mtx );
-      const auto tptr = local_txns.get<by_id>().find( tid );
-      return tptr != local_txns.end();
+      auto& id_idx = local_txns.get<by_id>();
+      if (auto tptr = id_idx.find(id); tptr != id_idx.end()) {
+         return tptr->connection_ids;
+      }
+      return {};
    }
 
    void dispatch_manager::expire_txns() {
+      fc::time_point now = time_point::now();
+
       size_t start_size = 0, end_size = 0;
-      fc::time_point_sec now{time_point::now()};
-
-      fc::unique_lock g( local_txns_mtx );
-      start_size = local_txns.size();
-      auto& old = local_txns.get<by_expiry>();
-      auto ex_lo = old.lower_bound( fc::time_point_sec( 0 ) );
-      auto ex_up = old.upper_bound( now );
-      old.erase( ex_lo, ex_up );
-      g.unlock();
+      flat_map<uint32_t, uint32_t> expired_trxs_for_connection;
+      {
+         fc::lock_guard g( local_txns_mtx );
+         start_size = local_txns.size();
+         auto& old = local_txns.get<by_expiry>();
+         auto ex_lo = old.lower_bound( fc::time_point_sec( 0 ) );
+         auto ex_up = old.upper_bound( fc::time_point_sec{now - fc::seconds(5)} ); // allow for some clock-skew
```

Comment:
I think it should be at least 10s (prob. a little more) to match what we use when we accept transactions.

Reply:
Increased to 15 seconds.

Reply:
https://github.com/AntelopeIO/spring/pull/1499/commits/9568000de1fda1ebdd8876f59a1e25b735f64bf7